Sync/BatchingClient.swift

```diff
@@ -0,0 +1,263 @@
+        }
+
+        // Need to deduce how big the data is we're sending across using this reducer
+        let sizeReducer: (total: ByteCount, record: Record<T>) -> ByteCount = { total, record in
+            return total + (self.serializeRecord(record) ?? "").utf8.count
+        }
+
+        let sizeOfRecords = records.reduce(0, combine: sizeReducer)
+
+        // Batch up all of the records and fail early if there is an issue
+        let batchingResult = batchesFromRecords(records)
+        guard let batches = batchingResult.successValue else {
+            let failure = batchingResult.failureValue!
+            log.debug("Unable to generate batches from records submitted to batch client: \(failure)")
+            return deferMaybe(failure)
+        }
+
+        // We have too many records for a single post so we'll either need to upload in a single batch or
+        // multiple batches
+        if records.count > config.maxTotalRecords || sizeOfRecords > config.maxTotalBytes {
+            return self.bunchUpload(batches, ifUnmodifiedSince: self.ifUnmodifiedSince, onCollectionUploaded: onCollectionUploaded)
+        }
+
+        if records.count > config.maxPostRecords || sizeOfRecords > config.maxPostBytes {
+            return self.batchUpload(batches, ifUnmodifiedSince: self.ifUnmodifiedSince, onCollectionUploaded: onCollectionUploaded)
+        }
+
+        // We can just do a single post instead of batching
+        log.debug("Batch fits within a single request. Submitting records in a single post.")
+        let lines = optFilter(records.map(self.serializeRecord))
+        return self.uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: nil)
+            >>== effect({ onCollectionUploaded($0.value) })
+            >>> succeed
+    }
+
+    public func bunchUpload(batch: [[String]], ifUnmodifiedSince: Timestamp?, onCollectionUploaded: (POSTResult -> Void)) -> Success {
+
+        // Need to break apart the provided batch into multiple batches (AKA a bunch) so we can fit them in batch calls
+        var mutBatch = batch
+
+        /* 
+         * When talking about a 'batch' and multiple 'batches' the word becomes very confusing. In this context,
+         *
+         * A 'batch' consists of N requests to the server beginning with a batch=true call and ends with a commit=true call.
+         * A 'bunch' is many of these calls.
+         * A partialBatch is one of the N requests made within a 'batch'.
+         *
+         * Here are some typealiases to make your life easier.
+         */
+
+        typealias Batch = [[String]]
+        typealias Bunch = [Batch]
+
+        var bunch = Bunch()
+
+        while !mutBatch.isEmpty {
+            var recordCount = 0
+            var size = 0
+            var smallerBatch = Batch()
+
+            while size <= config.maxTotalBytes && recordCount <= config.maxTotalRecords && !mutBatch.isEmpty {
+                let partialBatch = mutBatch.removeFirst()
```

Comment:
All of this nonsense becomes unnecessary if you do the work at the point of accumulation.

That is: when the caller calls `addRecord`, serialize the incoming record. If it fails, stop. If it succeeds, figure out how many bytes it'll be. If it's too big for one record, stop. Otherwise, accumulate it into a new or existing batch, and when a batch is full, accumulate that into a new or existing bunch.

Have tiers like this:

```
typealias UploadRecord = (guid: GUID, payload: String, size: Int)

class Payload {
    let config: InfoConfiguration
    var records: [UploadRecord] = []
    var bytes: Int = 0
    var count: Int = 0

    func take(record: UploadRecord) -> Bool {
        if (record.size + bytes) > config.maxPostBytes {
            return false
        }
        if count >= config.maxPostRecords {
            return false
        }
        …
        return true
}

class Batch {
     …
}

```

and so on. At each level you can keep track of bytes and counts. Note that streaming uploads (a good idea) involves keeping track of request, record, and byte counts so that you can choose when to start a new bunch.